lib/pf/iplog.pm

```diff
@@ -642,24 +650,63 @@ sub close {
     return (0);
 }
 
+sub rotate {
+    my $timer = pf::StatsD::Timer->new({sample_rate => 0.2});
+    my ( $window_seconds, $batch, $time_limit ) = @_;
+    my $logger = pf::log::get_logger();
+
+    $logger->debug("Calling rotate with window='$window_seconds' seconds, batch='$batch', timelimit='$time_limit'");
+    my $now = db_now();
+    my $start_time = time;
+    my $end_time;
+    my $rows_rotated = 0;
+
+    while (1) {
+        my $query;
+        my ( $rows_inserted, $rows_deleted );
+        pf::db::db_transaction_execute( sub{
+            $query = db_query_execute(IPLOG, $iplog_statements, 'iplog_rotate_insert_sql', $now, $window_seconds, $batch) || return (0);
```

Comment:
I have to say I find stored procedures to be less than obvious from an operations standpoint.
You have to remember that the database is doing some magic in the background, understand what it is doing, understand where the errors will be logged etc.

Straight up perl code is at least a single place to look for when it comes to debugging.